src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -2,143 +2,183 @@
+            {
+                string relativePath = Path.GetRelativePath(Path.GetDirectoryName(solutionFileFullPath), projectPath);
+                try
+                {
+                    AddProject(solution, relativePath, projectPath, solutionFolder);
+                }
+                catch (InvalidProjectFileException ex)
+                {
+                    Reporter.Error.WriteLine(string.Format(CommonLocalizableStrings.InvalidProjectWithExceptionMessage, projectPath, ex.Message));
+                }
+                catch (SolutionArgumentException ex) when (solution.FindProject(relativePath) != null || ex.Type == SolutionErrorType.DuplicateProjectName)
+                {
+                    Reporter.Output.WriteLine(CommonLocalizableStrings.SolutionAlreadyContainsProject, solutionFileFullPath, relativePath);
+                }
+            }
+            if (solution.SolutionProjects.Count > 1)
+            {
+                // https://stackoverflow.com/a/14714485
+                solution.RemoveProperties("HideSolutionNode");
+            }
+            await serializer.SaveAsync(solutionFileFullPath, solution, cancellationToken);
         }
 
-        private static bool IsPathInTreeRootedAtSolutionDirectory(string path)
+        private void AddProject(SolutionModel solution, string pathRelativeToSolutionFile, string fullPath, SolutionFolderModel solutionFolder)
         {
-            return !path.StartsWith("..");
-        }
+            // Open project instance to see if it is a valid project
+            ProjectRootElement projectRootElement = ProjectRootElement.Open(fullPath);
+            SolutionProjectModel project;
+            try
+            {
+                project = solution.AddProject(pathRelativeToSolutionFile, null, solutionFolder);
+            }
+            catch (SolutionArgumentException ex) when (ex.ParamName == "projectTypeName")
+            {
+                // If guid is not identified by vs-solutionpersistence, check in project element itself
+                var guid = projectRootElement.GetProjectTypeGuid();
+                if (string.IsNullOrEmpty(guid))
+                {
+                    Reporter.Error.WriteLine(CommonLocalizableStrings.UnsupportedProjectType, fullPath);
+                    return;
+                }
+                project = solution.AddProject(pathRelativeToSolutionFile, guid, solutionFolder);
+            }
+            // Generate intermediate solution folders
+            if (solutionFolder is null && !_inRoot)
+            {
+                var relativePathDirectory = Path.GetDirectoryName(pathRelativeToSolutionFile);
+                if (!string.IsNullOrEmpty(relativePathDirectory))
+                {
+                    SolutionFolderModel relativeSolutionFolder = solution.AddFolder(GetSolutionFolderPathWithForwardSlashes(relativePathDirectory));
+                    project.MoveToFolder(relativeSolutionFolder);
```

Comment:
Is it possible for there to be multiple folders in this hierarchy?

Reply:
You mean nested folders? If that is the case, yes, vs-solutionpersistence adds intermediate folders as needed